### PR TITLE
[WIP] Variable collector

### PR DESCRIPTION
**Pull request by dantleech**
This is an initial attempt at a variable collector (#222), which can be used by the `ReflectionMethod` to determine the variables that are used. The bulk of this code is concerned with resolving types so that:

```php
/**
 * @var Foobar
 */
public function something()
{
     $bar = $this->foobar->getSomeClass()->getFoo();
}
```

If the `getSomeClass` method has the return type `SomeClass` class and `SomeClass` a method `getFoo` that returns type `int`, then `$bar` will be resolved to `int`.

```php
$reflection = $reflector->reflect(SomeClass::class);
$vars = $reflection->getMethod('something')->getVariables();

echo $vars[0]->getName(); // bar
echo $vars[0]->getType(); // int
echo $vars[0]->getDeclaredAt(); // 6
```

The variable collector visitor can return locally assigned and parameter variables available in the AST it is traversing (regardless of scope).  In the case of getting the variables in a method it would suffice to traverse only the method AST.

This is really rough WIP atm and contains an unknown number of bugs, so before reviewing do you agree with the general approach?

TODO:

- [ ] Types from static methods/properties.
- [ ] ...

### CODE REVIEW

File: src/Reflection/ReflectionVariable.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private $declaredAt;
+
+    /**
+     * @var $type
+     */
+    private $type;
+
+    /**
+     * @param string $name
+     * @param int $declaredAt
```
Comment (Ocramius) [inline src/Reflection/ReflectionVariable.php]: Unsure about just having the line number.

File: src/Reflection/ReflectionVariable.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     * @param string $type
```
Comment (Ocramius) [inline src/Reflection/ReflectionVariable.php]: Doesn't say much - need some sort of more restrictive hint here

File: src/Reflection/ReflectionVariable.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        return $this->name;
+    }
+
+    public function getDeclaredAt() 
+    {
+        return $this->declaredAt;
+    }
+
+    /**
+     * Get a PhpDocumentor type object for this type
```
Comment (Ocramius) [inline src/Reflection/ReflectionVariable.php]: What if the type mutated during execution? Two variables with same name, but different type and declaration line, then?

File: src/Reflection/ReflectionVariable.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return (!$this->type instanceof Types\Object_);
+    }
+
+    /**
+     * Convert this string type to a string
+     *
+     * @see https://github.com/php/php-src/blob/master/ext/reflection/php_reflection.c#L2993
+     * @return string
+     */
+    public function __toString()
```
Comment (Ocramius) [inline src/Reflection/ReflectionVariable.php]: Wouldn't add this

File: src/Util/Visitor/VariableCollectionVisitor.php [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+                $types = $reflParam->getDocBlockTypes();
+                $type = count($types) ? ReflectionType::createFromType(reset($types), false) : null;
+            }
+
+            $this->methodParamTypes[$param->name] = $type;
+            $this->variables[] = ReflectionVariable::createFromName($param->name, $type, $param->getAttribute('startLine'));
+        }
+    }
+
+    private function processAssignation(Expr\Assign $node)
```
Comment (Ocramius) [inline src/Util/Visitor/VariableCollectionVisitor.php]: Just `processAssign`

**Comment by Ocramius** [comment]
Approach looking sound to me. Loads of testing and refactoring to be done though

File: src/Reflection/ReflectionVariable.php (hunk 1 above)
Comment (dantleech) [inline src/Reflection/ReflectionVariable.php]: Well, ideally there would be the range of (characters?) that the variable is available for, but the PHP parser just gives us line numbers ..

File: src/Reflection/ReflectionVariable.php (hunk 3 above)
Comment (dantleech) [inline src/Reflection/ReflectionVariable.php]: We return all declarations, so you could get two with the same name, but covering different parts of the code.

File: src/Reflection/ReflectionVariable.php (hunk 4 above)
Comment (dantleech) [inline src/Reflection/ReflectionVariable.php]: I found it useful for testing, but mainly I added it to be consistent with the other `Reflection*` classes.

File: src/Reflection/ReflectionVariable.php (hunk 2 above)
Comment (dantleech) [inline src/Reflection/ReflectionVariable.php]: Its actually a `ReflectionType` ...

File: src/Reflection/ReflectionVariable.php (hunk 4 above)
Comment (Ocramius) [inline src/Reflection/ReflectionVariable.php]: Unless it actually is required API, I wouldn't add it. `__toString()` is often abused, so let's not give users more rope to hang themselves :)

File: .travis.yml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 language: php
 
 php:
-  - 5.6
+    #  - 5.6
```
Comment (dantleech) [inline .travis.yml]: Have switched to using PHP7 at least for development - guessing that you are going to switch to PHP7 in the near-future anyway?

File: src/Util/Visitor/VariableCollectionVisitor.php [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+    private function createReflectionTypeFromDocType(Type $type): ReflectionType
+    {
+        return ReflectionType::createFromType($type, false);
+    }
+
+    private function createUnknownReflectionType()
+    {
+        return $this->createReflectionTypeFromString('mixed', false);
```
Comment (dantleech) [inline src/Util/Visitor/VariableCollectionVisitor.php]: What should we do when the type cannot be determined? Currently returning `mixed` - should we return the type as `null`?

Comment (Ocramius) [inline src/Util/Visitor/VariableCollectionVisitor.php]: Should be `null`, IMO. Adding `mixed` is a very specific thing, plus nothing prevents a `class mixed {}` to exist.

File: src/Util/Visitor/VariableCollectionVisitor.php [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        // if this is a function call, try and instantiate the runtime native
+        // \ReflectionFunction, if it is not internal then ignore it as we
+        // cannot guarantee that we are running in the same process as the code
+        // we are analyzing.
+        //
+        // TODO: This is no positive test case for this ... which PHP internal functions
+        //       actually have a return type??
+        if ($expr instanceof Expr\FuncCall) {
+            $func = (string) $expr->name;
+            $reflection = new \ReflectionFunction($func);
```
Comment (Ocramius) [inline src/Util/Visitor/VariableCollectionVisitor.php]: You can stop at the return type declared on the function btw - no need to introspect.

**Comment by dantleech** [comment]
At the moment this PR scope is about collecting "available" variables (#222), currently including the parameters given in a method call. If taken further this means we should also return `$this` as a variable, and further still, globals, which we cannot do.

For my specifc use case, I need the "available" variables (albeit without the globals). 

Do you think this functionality be limited to variables *assigned* within a scope? Or should we return the func/method param variables also?

**Comment by MarkRedeman** [comment]
@dantleech I think there are still a couple of variable definitions that you're missing in this PR.

- Variable definition through a `foreach` statement. (felixfbecker/php-language-server/#200)
- Pass by reference (felixfbecker/php-language-server#205)
- Assignment trough lists (felixfbecker/php-language-server#170)
- Array destructuring (PHP 7.1+, https://wiki.php.net/rfc/short_list_syntax)

**Comment by dantleech** [comment]
> I think there are still a couple of variable definitions that you're missing in this PR.

indeed, it is far from being complete.

> Variable definition through a foreach statement. 

How? PHPStorm would rely on an annotation above the foreach loop here I think (?)

**Comment by MarkRedeman** [comment]
> indeed, it is far from being complete.

I understand, I mostly wanted to make you aware of those use cases :).

> How? PHPStorm would rely on an annotation above the foreach loop here I think (?)

That's probably going to be a tricky one.. You might be able to check whether the type of the iterable. So that if you're iterating over an `Iterator` then you can check the return type of `current`.
If the iterable is an `IteratorAggregate` then this would require some recursive checks.
For arrays I guess the best use case would be to guess their types from the docblocks.

Btw the use cases I mentioned were mostly about finding the definitions of a variable, not necessarily including their types (though it would be very useful).

**Comment by asgrim** [comment]
@dantleech note: the core namespace is now `\Roave\BetterReflection` so this PR will need updating. No need to provide BC shim class entries for this PR as it's a new feature anyway :)

**Comment by TomasVotruba** [comment]
@dantleech Hey, thanks for this awesome PR. I need exactly this feature, just without reflection and based on `nikic\php-parser`, so I might take a lot from it.

Do you expect to finish this now on BR 2.0? I could help if you need

**Comment by dantleech** [comment]
@TomasVotruba I don't anticipate working on this, I already started [another project](https://github.com/phpactor/worse-reflection) which does this. The problem is that it is a huge amount of work with lots of edge cases. My implementation works "OK" for the purpose (auto completion).

**Comment by TomasVotruba** [comment]
@dantleech Oh, I found out your package [WorseReflection](https://github.com/phpactor/worse-reflection) where you've probably implemented this.

Where could I find there this feature? I looked to [ReflectionMethod](https://github.com/phpactor/worse-reflection/blob/master/lib/Bridge/TolerantParser/Reflection/ReflectionMethod.php) but could not find it.

**Comment by dantleech** [comment]
The entry point is `Reflector::reflectOffset(): ReflectionOffset` (`$offset->frame()` returns the frame).

Most of the code is [here](https://github.com/phpactor/worse-reflection/blob/master/lib/Core/Inference/SymbolInformationResolver.php) and close by.

**Comment by TomasVotruba** [comment]
@dantleech Not sure why, but I didn't see your comments on mobile, that explains my comments :).
Neither I got notified about your last comment. Thanks a lot for that, I'll look on it.

How are you satisfied with the `"microsoft/tolerant-php-parser"`?

Any useful code parts that you think might be useful for me?

**Comment by dantleech** [comment]
In general I am happy with tolerant parser. It's designed for IDE usage, the API is richer than `nikic/php-parser`. For example `$node->getFirstAncestor(FunctionLike::class, ClassLike::class)`, `$node->getDescendantAtOffset(1234)`. It also will resolve the fully qualified names of things, so `$node->getFirstAncestor(ClassLike::class)->getFullyQualifiedName()` etc.

**Comment by TomasVotruba** [comment]
@dantleech Thanks, I'll look on these features. 
I managed to implement most of them via custom `NodeVisitors` already.

Just off-topic: It looks like you have found another solution, so maybe this PR can be closed. I thouth it's WIP and expected it might be finished, but now I see there is complete another way you took. What do you think?

**Comment by dantleech** [comment]
The approach is basically the same - the only reason I created WorseReflection was because it's the core domain of Phpactor and I had lots of requirements. I think the WR implementation could be used as a reference for implementing this feature in the future (I also think the PHP Language Server would be worth a look in this respect).

**Comment by TomasVotruba** [comment]
In case anyone would be looking for `getFirstAncestor` in nikic/php-parser, here it is: https://github.com/RectorPHP/Rector/blob/d59ffe27e33c9a63df5b9e9cf2f0a432e4f4c401/packages/NodeTraverserQueue/src/BetterNodeFinder.php

**Pull request closed without merging**